// jotai/umd/babel/plugin-react-refresh.development.js
(function (global, factory) {
  typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@babel/template')) :
  typeof define === 'function' && define.amd ? define(['exports', '@babel/template'], factory) :
  (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global["jotaiPlugin-react-refresh"] = {}, global._templateBuilder));
})(this, (function (exports, _templateBuilder) { 'use strict';

  function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

  var _templateBuilder__default = /*#__PURE__*/_interopDefaultLegacy(_templateBuilder);

  function isAtom(t, callee) {
    if (t.isIdentifier(callee) && atomFunctionNames.includes(callee.name)) {
      return true;
    }

    if (t.isMemberExpression(callee)) {
      var property = callee.property;

      if (t.isIdentifier(property) && atomFunctionNames.includes(property.name)) {
        return true;
      }
    }

    return false;
  }
  var atomFunctionNames = ['abortableAtom', 'atom', 'atomFamily', 'atomWithDefault', 'atomWithHash', 'atomWithImmer', 'atomWithInfiniteQuery', 'atomWithMachine', 'atomWithMutation', 'atomWithObservable', 'atomWithProxy', 'atomWithQuery', 'atomWithReducer', 'atomWithReset', 'atomWithSubscription', 'atomWithStorage', 'atomWithStore', 'freezeAtom', 'loadable', 'selectAtom', 'splitAtom'];

  var templateBuilder = _templateBuilder__default["default"].default || _templateBuilder__default["default"];
  function reactRefreshPlugin(_ref) {
    var t = _ref.types;
    return {
      pre: function pre(_ref2) {
        var opts = _ref2.opts;

        if (!opts.filename) {
          throw new Error('Filename must be available');
        }
      },
      visitor: {
        Program: {
          exit: function exit(path) {
            var jotaiAtomCache = templateBuilder("\n          globalThis.jotaiAtomCache = globalThis.jotaiAtomCache || {\n            cache: new Map(),\n            get(name, inst) { \n              if (this.cache.has(name)) {\n                return this.cache.get(name)\n              }\n              this.cache.set(name, inst)\n              return inst\n            },\n          }")();
            path.unshiftContainer('body', jotaiAtomCache);
          }
        },
        ExportDefaultDeclaration: function ExportDefaultDeclaration(nodePath, state) {
          var node = nodePath.node;

          if (t.isCallExpression(node.declaration) && isAtom(t, node.declaration.callee)) {
            var filename = state.filename || 'unknown';
            var atomKey = filename + "/defaultExport";
            var buildExport = templateBuilder("export default globalThis.jotaiAtomCache.get(%%atomKey%%, %%atom%%)");
            var ast = buildExport({
              atomKey: t.stringLiteral(atomKey),
              atom: node.declaration
            });
            nodePath.replaceWith(ast);
          }
        },
        VariableDeclarator: function VariableDeclarator(nodePath, state) {
          var _nodePath$parentPath$, _nodePath$parentPath$2;

          if (t.isIdentifier(nodePath.node.id) && t.isCallExpression(nodePath.node.init) && isAtom(t, nodePath.node.init.callee) && ((_nodePath$parentPath$ = nodePath.parentPath.parentPath) != null && _nodePath$parentPath$.isProgram() || (_nodePath$parentPath$2 = nodePath.parentPath.parentPath) != null && _nodePath$parentPath$2.isExportNamedDeclaration())) {
            var filename = state.filename || 'unknown';
            var atomKey = filename + "/" + nodePath.node.id.name;
            var buildAtomDeclaration = templateBuilder("const %%atomIdentifier%% = globalThis.jotaiAtomCache.get(%%atomKey%%, %%atom%%)");
            var ast = buildAtomDeclaration({
              atomIdentifier: t.identifier(nodePath.node.id.name),
              atomKey: t.stringLiteral(atomKey),
              atom: nodePath.node.init
            });
            nodePath.parentPath.replaceWith(ast);
          }
        }
      }
    };
  }

  exports["default"] = reactRefreshPlugin;

  Object.defineProperty(exports, '__esModule', { value: true });

}));
