// @expo/webpack-config/webpack/addons/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.withPlatformSourceMaps = exports.withEntry = exports.withNodeMocks = exports.withDevServer = exports.withAlias = exports.withOptimizations = exports.withUnimodules = void 0;
var withUnimodules_1 = require("./withUnimodules");
Object.defineProperty(exports, "withUnimodules", { enumerable: true, get: function () { return __importDefault(withUnimodules_1).default; } });
var withOptimizations_1 = require("./withOptimizations");
Object.defineProperty(exports, "withOptimizations", { enumerable: true, get: function () { return __importDefault(withOptimizations_1).default; } });
var withAlias_1 = require("./withAlias");
Object.defineProperty(exports, "withAlias", { enumerable: true, get: function () { return __importDefault(withAlias_1).default; } });
var withDevServer_1 = require("./withDevServer");
Object.defineProperty(exports, "withDevServer", { enumerable: true, get: function () { return __importDefault(withDevServer_1).default; } });
var withNodeMocks_1 = require("./withNodeMocks");
Object.defineProperty(exports, "withNodeMocks", { enumerable: true, get: function () { return __importDefault(withNodeMocks_1).default; } });
var withEntry_1 = require("./withEntry");
Object.defineProperty(exports, "withEntry", { enumerable: true, get: function () { return __importDefault(withEntry_1).default; } });
var withSourceMaps_1 = require("./withSourceMaps");
Object.defineProperty(exports, "withPlatformSourceMaps", { enumerable: true, get: function () { return __importDefault(withSourceMaps_1).default; } });
