// jotai/urql.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var core = require('@urql/core');
var jotai = require('jotai');
var wonka = require('wonka');

var DEFAULT_URL = function () {
  try {
    return process.env.JOTAI_URQL_DEFAULT_URL;
  } catch (_unused) {
    return undefined;
  }
}() || '/graphql';

var clientAtom = jotai.atom(core.createClient({
  url: DEFAULT_URL
}));

function _regeneratorRuntime() {
  /*! regenerator-runtime -- Copyright (c) 2014-present, Facebook, Inc. -- license (MIT): https://github.com/facebook/regenerator/blob/main/LICENSE */

  _regeneratorRuntime = function () {
    return exports;
  };

  var exports = {},
      Op = Object.prototype,
      hasOwn = Op.hasOwnProperty,
      $Symbol = "function" == typeof Symbol ? Symbol : {},
      iteratorSymbol = $Symbol.iterator || "@@iterator",
      asyncIteratorSymbol = $Symbol.asyncIterator || "@@asyncIterator",
      toStringTagSymbol = $Symbol.toStringTag || "@@toStringTag";

  function define(obj, key, value) {
    return Object.defineProperty(obj, key, {
      value: value,
      enumerable: !0,
      configurable: !0,
      writable: !0
    }), obj[key];
  }

  try {
    define({}, "");
  } catch (err) {
    define = function (obj, key, value) {
      return obj[key] = value;
    };
  }

  function wrap(innerFn, outerFn, self, tryLocsList) {
    var protoGenerator = outerFn && outerFn.prototype instanceof Generator ? outerFn : Generator,
        generator = Object.create(protoGenerator.prototype),
        context = new Context(tryLocsList || []);
    return generator._invoke = function (innerFn, self, context) {
      var state = "suspendedStart";
      return function (method, arg) {
        if ("executing" === state) throw new Error("Generator is already running");

        if ("completed" === state) {
          if ("throw" === method) throw arg;
          return doneResult();
        }

        for (context.method = method, context.arg = arg;;) {
          var delegate = context.delegate;

          if (delegate) {
            var delegateResult = maybeInvokeDelegate(delegate, context);

            if (delegateResult) {
              if (delegateResult === ContinueSentinel) continue;
              return delegateResult;
            }
          }

          if ("next" === context.method) context.sent = context._sent = context.arg;else if ("throw" === context.method) {
            if ("suspendedStart" === state) throw state = "completed", context.arg;
            context.dispatchException(context.arg);
          } else "return" === context.method && context.abrupt("return", context.arg);
          state = "executing";
          var record = tryCatch(innerFn, self, context);

          if ("normal" === record.type) {
            if (state = context.done ? "completed" : "suspendedYield", record.arg === ContinueSentinel) continue;
            return {
              value: record.arg,
              done: context.done
            };
          }

          "throw" === record.type && (state = "completed", context.method = "throw", context.arg = record.arg);
        }
      };
    }(innerFn, self, context), generator;
  }

  function tryCatch(fn, obj, arg) {
    try {
      return {
        type: "normal",
        arg: fn.call(obj, arg)
      };
    } catch (err) {
      return {
        type: "throw",
        arg: err
      };
    }
  }

  exports.wrap = wrap;
  var ContinueSentinel = {};

  function Generator() {}

  function GeneratorFunction() {}

  function GeneratorFunctionPrototype() {}

  var IteratorPrototype = {};
  define(IteratorPrototype, iteratorSymbol, function () {
    return this;
  });
  var getProto = Object.getPrototypeOf,
      NativeIteratorPrototype = getProto && getProto(getProto(values([])));
  NativeIteratorPrototype && NativeIteratorPrototype !== Op && hasOwn.call(NativeIteratorPrototype, iteratorSymbol) && (IteratorPrototype = NativeIteratorPrototype);
  var Gp = GeneratorFunctionPrototype.prototype = Generator.prototype = Object.create(IteratorPrototype);

  function defineIteratorMethods(prototype) {
    ["next", "throw", "return"].forEach(function (method) {
      define(prototype, method, function (arg) {
        return this._invoke(method, arg);
      });
    });
  }

  function AsyncIterator(generator, PromiseImpl) {
    function invoke(method, arg, resolve, reject) {
      var record = tryCatch(generator[method], generator, arg);

      if ("throw" !== record.type) {
        var result = record.arg,
            value = result.value;
        return value && "object" == typeof value && hasOwn.call(value, "__await") ? PromiseImpl.resolve(value.__await).then(function (value) {
          invoke("next", value, resolve, reject);
        }, function (err) {
          invoke("throw", err, resolve, reject);
        }) : PromiseImpl.resolve(value).then(function (unwrapped) {
          result.value = unwrapped, resolve(result);
        }, function (error) {
          return invoke("throw", error, resolve, reject);
        });
      }

      reject(record.arg);
    }

    var previousPromise;

    this._invoke = function (method, arg) {
      function callInvokeWithMethodAndArg() {
        return new PromiseImpl(function (resolve, reject) {
          invoke(method, arg, resolve, reject);
        });
      }

      return previousPromise = previousPromise ? previousPromise.then(callInvokeWithMethodAndArg, callInvokeWithMethodAndArg) : callInvokeWithMethodAndArg();
    };
  }

  function maybeInvokeDelegate(delegate, context) {
    var method = delegate.iterator[context.method];

    if (undefined === method) {
      if (context.delegate = null, "throw" === context.method) {
        if (delegate.iterator.return && (context.method = "return", context.arg = undefined, maybeInvokeDelegate(delegate, context), "throw" === context.method)) return ContinueSentinel;
        context.method = "throw", context.arg = new TypeError("The iterator does not provide a 'throw' method");
      }

      return ContinueSentinel;
    }

    var record = tryCatch(method, delegate.iterator, context.arg);
    if ("throw" === record.type) return context.method = "throw", context.arg = record.arg, context.delegate = null, ContinueSentinel;
    var info = record.arg;
    return info ? info.done ? (context[delegate.resultName] = info.value, context.next = delegate.nextLoc, "return" !== context.method && (context.method = "next", context.arg = undefined), context.delegate = null, ContinueSentinel) : info : (context.method = "throw", context.arg = new TypeError("iterator result is not an object"), context.delegate = null, ContinueSentinel);
  }

  function pushTryEntry(locs) {
    var entry = {
      tryLoc: locs[0]
    };
    1 in locs && (entry.catchLoc = locs[1]), 2 in locs && (entry.finallyLoc = locs[2], entry.afterLoc = locs[3]), this.tryEntries.push(entry);
  }

  function resetTryEntry(entry) {
    var record = entry.completion || {};
    record.type = "normal", delete record.arg, entry.completion = record;
  }

  function Context(tryLocsList) {
    this.tryEntries = [{
      tryLoc: "root"
    }], tryLocsList.forEach(pushTryEntry, this), this.reset(!0);
  }

  function values(iterable) {
    if (iterable) {
      var iteratorMethod = iterable[iteratorSymbol];
      if (iteratorMethod) return iteratorMethod.call(iterable);
      if ("function" == typeof iterable.next) return iterable;

      if (!isNaN(iterable.length)) {
        var i = -1,
            next = function next() {
          for (; ++i < iterable.length;) if (hasOwn.call(iterable, i)) return next.value = iterable[i], next.done = !1, next;

          return next.value = undefined, next.done = !0, next;
        };

        return next.next = next;
      }
    }

    return {
      next: doneResult
    };
  }

  function doneResult() {
    return {
      value: undefined,
      done: !0
    };
  }

  return GeneratorFunction.prototype = GeneratorFunctionPrototype, define(Gp, "constructor", GeneratorFunctionPrototype), define(GeneratorFunctionPrototype, "constructor", GeneratorFunction), GeneratorFunction.displayName = define(GeneratorFunctionPrototype, toStringTagSymbol, "GeneratorFunction"), exports.isGeneratorFunction = function (genFun) {
    var ctor = "function" == typeof genFun && genFun.constructor;
    return !!ctor && (ctor === GeneratorFunction || "GeneratorFunction" === (ctor.displayName || ctor.name));
  }, exports.mark = function (genFun) {
    return Object.setPrototypeOf ? Object.setPrototypeOf(genFun, GeneratorFunctionPrototype) : (genFun.__proto__ = GeneratorFunctionPrototype, define(genFun, toStringTagSymbol, "GeneratorFunction")), genFun.prototype = Object.create(Gp), genFun;
  }, exports.awrap = function (arg) {
    return {
      __await: arg
    };
  }, defineIteratorMethods(AsyncIterator.prototype), define(AsyncIterator.prototype, asyncIteratorSymbol, function () {
    return this;
  }), exports.AsyncIterator = AsyncIterator, exports.async = function (innerFn, outerFn, self, tryLocsList, PromiseImpl) {
    void 0 === PromiseImpl && (PromiseImpl = Promise);
    var iter = new AsyncIterator(wrap(innerFn, outerFn, self, tryLocsList), PromiseImpl);
    return exports.isGeneratorFunction(outerFn) ? iter : iter.next().then(function (result) {
      return result.done ? result.value : iter.next();
    });
  }, defineIteratorMethods(Gp), define(Gp, toStringTagSymbol, "Generator"), define(Gp, iteratorSymbol, function () {
    return this;
  }), define(Gp, "toString", function () {
    return "[object Generator]";
  }), exports.keys = function (object) {
    var keys = [];

    for (var key in object) keys.push(key);

    return keys.reverse(), function next() {
      for (; keys.length;) {
        var key = keys.pop();
        if (key in object) return next.value = key, next.done = !1, next;
      }

      return next.done = !0, next;
    };
  }, exports.values = values, Context.prototype = {
    constructor: Context,
    reset: function (skipTempReset) {
      if (this.prev = 0, this.next = 0, this.sent = this._sent = undefined, this.done = !1, this.delegate = null, this.method = "next", this.arg = undefined, this.tryEntries.forEach(resetTryEntry), !skipTempReset) for (var name in this) "t" === name.charAt(0) && hasOwn.call(this, name) && !isNaN(+name.slice(1)) && (this[name] = undefined);
    },
    stop: function () {
      this.done = !0;
      var rootRecord = this.tryEntries[0].completion;
      if ("throw" === rootRecord.type) throw rootRecord.arg;
      return this.rval;
    },
    dispatchException: function (exception) {
      if (this.done) throw exception;
      var context = this;

      function handle(loc, caught) {
        return record.type = "throw", record.arg = exception, context.next = loc, caught && (context.method = "next", context.arg = undefined), !!caught;
      }

      for (var i = this.tryEntries.length - 1; i >= 0; --i) {
        var entry = this.tryEntries[i],
            record = entry.completion;
        if ("root" === entry.tryLoc) return handle("end");

        if (entry.tryLoc <= this.prev) {
          var hasCatch = hasOwn.call(entry, "catchLoc"),
              hasFinally = hasOwn.call(entry, "finallyLoc");

          if (hasCatch && hasFinally) {
            if (this.prev < entry.catchLoc) return handle(entry.catchLoc, !0);
            if (this.prev < entry.finallyLoc) return handle(entry.finallyLoc);
          } else if (hasCatch) {
            if (this.prev < entry.catchLoc) return handle(entry.catchLoc, !0);
          } else {
            if (!hasFinally) throw new Error("try statement without catch or finally");
            if (this.prev < entry.finallyLoc) return handle(entry.finallyLoc);
          }
        }
      }
    },
    abrupt: function (type, arg) {
      for (var i = this.tryEntries.length - 1; i >= 0; --i) {
        var entry = this.tryEntries[i];

        if (entry.tryLoc <= this.prev && hasOwn.call(entry, "finallyLoc") && this.prev < entry.finallyLoc) {
          var finallyEntry = entry;
          break;
        }
      }

      finallyEntry && ("break" === type || "continue" === type) && finallyEntry.tryLoc <= arg && arg <= finallyEntry.finallyLoc && (finallyEntry = null);
      var record = finallyEntry ? finallyEntry.completion : {};
      return record.type = type, record.arg = arg, finallyEntry ? (this.method = "next", this.next = finallyEntry.finallyLoc, ContinueSentinel) : this.complete(record);
    },
    complete: function (record, afterLoc) {
      if ("throw" === record.type) throw record.arg;
      return "break" === record.type || "continue" === record.type ? this.next = record.arg : "return" === record.type ? (this.rval = this.arg = record.arg, this.method = "return", this.next = "end") : "normal" === record.type && afterLoc && (this.next = afterLoc), ContinueSentinel;
    },
    finish: function (finallyLoc) {
      for (var i = this.tryEntries.length - 1; i >= 0; --i) {
        var entry = this.tryEntries[i];
        if (entry.finallyLoc === finallyLoc) return this.complete(entry.completion, entry.afterLoc), resetTryEntry(entry), ContinueSentinel;
      }
    },
    catch: function (tryLoc) {
      for (var i = this.tryEntries.length - 1; i >= 0; --i) {
        var entry = this.tryEntries[i];

        if (entry.tryLoc === tryLoc) {
          var record = entry.completion;

          if ("throw" === record.type) {
            var thrown = record.arg;
            resetTryEntry(entry);
          }

          return thrown;
        }
      }

      throw new Error("illegal catch attempt");
    },
    delegateYield: function (iterable, resultName, nextLoc) {
      return this.delegate = {
        iterator: values(iterable),
        resultName: resultName,
        nextLoc: nextLoc
      }, "next" === this.method && (this.arg = undefined), ContinueSentinel;
    }
  }, exports;
}

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) {
  try {
    var info = gen[key](arg);
    var value = info.value;
  } catch (error) {
    reject(error);
    return;
  }

  if (info.done) {
    resolve(value);
  } else {
    Promise.resolve(value).then(_next, _throw);
  }
}

function _asyncToGenerator(fn) {
  return function () {
    var self = this,
        args = arguments;
    return new Promise(function (resolve, reject) {
      var gen = fn.apply(self, args);

      function _next(value) {
        asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value);
      }

      function _throw(err) {
        asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err);
      }

      _next(undefined);
    });
  };
}

function _extends() {
  _extends = Object.assign ? Object.assign.bind() : function (target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];

      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }

    return target;
  };
  return _extends.apply(this, arguments);
}

var isOperationResultWithData$1 = function isOperationResultWithData(result) {
  return 'data' in result && !result.error;
};

function atomWithQuery(createQueryArgs, getClient) {
  if (getClient === void 0) {
    getClient = function getClient(get) {
      return get(clientAtom);
    };
  }

  var queryResultAtom = jotai.atom(function (get) {
    var args = createQueryArgs(get);

    if (args.pause) {
      return null;
    }

    var client = getClient(get);
    var resolve = null;

    var makePending = function makePending() {
      return new Promise(function (r) {
        resolve = r;
      });
    };

    var resultAtom = jotai.atom(makePending());
    var setResult = null;

    var listener = function listener(result) {
      if (!resolve && !setResult) {
        throw new Error('setting result without mount');
      }

      if (resolve) {
        resolve(result);
        resolve = null;
      }

      if (setResult) {
        setResult(result);
      }
    };

    var subscription = null;
    var timer;

    var startQuery = function startQuery(opts) {
      if (subscription) {
        clearTimeout(timer);
        subscription.unsubscribe();
      }

      subscription = wonka.pipe(client.query(args.query, args.variables, _extends({}, args.requestPolicy && {
        requestPolicy: args.requestPolicy
      }, args.context, opts)), wonka.subscribe(listener));

      if (!setResult) {
        timer = setTimeout(function () {
          if (subscription) {
            subscription.unsubscribe();
            subscription = null;
          }
        }, 1000);
      }
    };

    startQuery();

    resultAtom.onMount = function (update) {
      setResult = update;

      if (subscription) {
        clearTimeout(timer);
      } else {
        startQuery();
      }

      return function () {
        setResult = null;

        if (subscription) {
          subscription.unsubscribe();
          subscription = null;
        }
      };
    };

    return {
      resultAtom: resultAtom,
      makePending: makePending,
      startQuery: startQuery
    };
  });
  var queryAtom = jotai.atom(function (get) {
    var queryResult = get(queryResultAtom);

    if (!queryResult) {
      return null;
    }

    var resultAtom = queryResult.resultAtom;
    var result = get(resultAtom);

    if (!isOperationResultWithData$1(result)) {
      throw result.error;
    }

    return result;
  }, function (get, set, action) {
    if (action.type === 'reexecute') {
      console.warn('DEPRECATED [atomWithQuery] use refetch instead of reexecute');
      action.type = 'refetch';
    }

    switch (action.type) {
      case 'refetch':
        {
          var queryResult = get(queryResultAtom);

          if (!queryResult) {
            throw new Error('query is paused');
          }

          var resultAtom = queryResult.resultAtom,
              makePending = queryResult.makePending,
              startQuery = queryResult.startQuery;
          set(resultAtom, makePending());
          startQuery(action.opts);
          return;
        }
    }
  });
  return queryAtom;
}

function atomWithMutation(createQuery, getClient) {
  if (getClient === void 0) {
    getClient = function getClient(get) {
      return get(clientAtom);
    };
  }

  var operationResultAtom = jotai.atom(new Promise(function () {}));
  var queryResultAtom = jotai.atom(function (get) {
    return get(operationResultAtom);
  }, function () {
    var _ref = _asyncToGenerator(_regeneratorRuntime().mark(function _callee(get, set, action) {
      var client, query;
      return _regeneratorRuntime().wrap(function _callee$(_context) {
        while (1) {
          switch (_context.prev = _context.next) {
            case 0:
              set(operationResultAtom, new Promise(function () {}));
              client = getClient(get);
              query = createQuery(get);
              return _context.abrupt("return", client.mutation(query, action.variables, action.context).toPromise().then(function (result) {
                action.callback == null ? void 0 : action.callback(result);

                if (result.error) {
                  throw result.error;
                }

                set(operationResultAtom, result);
              }));

            case 4:
            case "end":
              return _context.stop();
          }
        }
      }, _callee);
    }));

    return function (_x, _x2, _x3) {
      return _ref.apply(this, arguments);
    };
  }());
  return queryResultAtom;
}

var isOperationResultWithData = function isOperationResultWithData(result) {
  return 'data' in result && !result.error;
};

function atomWithSubscription(createSubscriptionArgs, getClient) {
  if (getClient === void 0) {
    getClient = function getClient(get) {
      return get(clientAtom);
    };
  }

  var queryResultAtom = jotai.atom(function (get) {
    var args = createSubscriptionArgs(get);

    if (args.pause) {
      return null;
    }

    var client = getClient(get);
    var resolve = null;

    var makePending = function makePending() {
      return new Promise(function (r) {
        resolve = r;
      });
    };

    var resultAtom = jotai.atom(makePending());
    var setResult = null;

    var listener = function listener(result) {
      if (resolve) {
        resolve(result);
        resolve = null;
      }

      if (setResult) {
        setResult(result);
      }
    };

    var subscription = null;
    var timer;

    var startSub = function startSub() {
      if (subscription) {
        clearTimeout(timer);
        subscription.unsubscribe();
      }

      subscription = wonka.pipe(client.subscription(args.query, args.variables, args.context), wonka.subscribe(listener));

      if (!setResult) {
        timer = setTimeout(function () {
          if (subscription) {
            subscription.unsubscribe();
            subscription = null;
          }
        }, 1000);
      }
    };

    startSub();

    resultAtom.onMount = function (update) {
      setResult = update;

      if (subscription) {
        clearTimeout(timer);
      } else {
        startSub();
      }

      return function () {
        setResult = null;

        if (subscription) {
          subscription.unsubscribe();
          subscription = null;
        }
      };
    };

    return {
      resultAtom: resultAtom,
      makePending: makePending,
      startSub: startSub
    };
  });
  var queryAtom = jotai.atom(function (get) {
    var queryResult = get(queryResultAtom);

    if (!queryResult) {
      return null;
    }

    var resultAtom = queryResult.resultAtom;
    var result = get(resultAtom);

    if (!isOperationResultWithData(result)) {
      throw result.error;
    }

    return result;
  }, function (get, set, action) {
    switch (action.type) {
      case 'refetch':
        {
          var queryResult = get(queryResultAtom);

          if (!queryResult) {
            throw new Error('query is paused');
          }

          var resultAtom = queryResult.resultAtom,
              makePending = queryResult.makePending,
              startSub = queryResult.startSub;
          set(resultAtom, makePending());
          startSub();
          return;
        }
    }
  });
  return queryAtom;
}

exports.atomWithMutation = atomWithMutation;
exports.atomWithQuery = atomWithQuery;
exports.atomWithSubscription = atomWithSubscription;
exports.clientAtom = clientAtom;
