// @testing-library/react-native/build/queries/a11yValue.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.bindByA11yValueQueries = void 0;

var _matchObjectProp = require("../helpers/matchers/matchObjectProp");

var _makeQueries = require("./makeQueries");

const queryAllByA11yValue = instance => function queryAllByA11yValueFn(value) {
  return instance.findAll(node => typeof node.type === 'string' && (0, _matchObjectProp.matchObjectProp)(node.props.accessibilityValue, value));
};

const getMultipleError = value => `Found multiple elements with accessibilityValue: ${JSON.stringify(value)} `;

const getMissingError = value => `Unable to find an element with accessibilityValue: ${JSON.stringify(value)}`;

const {
  getBy,
  getAllBy,
  queryBy,
  queryAllBy,
  findBy,
  findAllBy
} = (0, _makeQueries.makeQueries)(queryAllByA11yValue, getMissingError, getMultipleError);

const bindByA11yValueQueries = instance => {
  const getByA11yValue = getBy(instance);
  const getAllByA11yValue = getAllBy(instance);
  const queryByA11yValue = queryBy(instance);
  const queryAllByA11yValue = queryAllBy(instance);
  const findByA11yValue = findBy(instance);
  const findAllByA11yValue = findAllBy(instance);
  return {
    getByA11yValue,
    getAllByA11yValue,
    queryByA11yValue,
    queryAllByA11yValue,
    findByA11yValue,
    findAllByA11yValue,
    getByAccessibilityValue: getByA11yValue,
    getAllByAccessibilityValue: getAllByA11yValue,
    queryByAccessibilityValue: queryByA11yValue,
    queryAllByAccessibilityValue: queryAllByA11yValue,
    findByAccessibilityValue: findByA11yValue,
    findAllByAccessibilityValue: findAllByA11yValue
  };
};

exports.bindByA11yValueQueries = bindByA11yValueQueries;
