// @testing-library/react-native/build/queries/a11yState.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.bindByA11yStateQueries = void 0;

var _matchObjectProp = require("../helpers/matchers/matchObjectProp");

var _makeQueries = require("./makeQueries");

const queryAllByA11yState = instance => function queryAllByA11yStateFn(state) {
  return instance.findAll(node => typeof node.type === 'string' && (0, _matchObjectProp.matchObjectProp)(node.props.accessibilityState, state));
};

const getMultipleError = state => `Found multiple elements with accessibilityState: ${JSON.stringify(state)}`;

const getMissingError = state => `Unable to find an element with accessibilityState: ${JSON.stringify(state)}`;

const {
  getBy,
  getAllBy,
  queryBy,
  queryAllBy,
  findBy,
  findAllBy
} = (0, _makeQueries.makeQueries)(queryAllByA11yState, getMissingError, getMultipleError);

const bindByA11yStateQueries = instance => {
  const getByA11yState = getBy(instance);
  const getAllByA11yState = getAllBy(instance);
  const queryByA11yState = queryBy(instance);
  const queryAllByA11yState = queryAllBy(instance);
  const findByA11yState = findBy(instance);
  const findAllByA11yState = findAllBy(instance);
  return {
    getByA11yState,
    getAllByA11yState,
    queryByA11yState,
    queryAllByA11yState,
    findByA11yState,
    findAllByA11yState,
    getByAccessibilityState: getByA11yState,
    getAllByAccessibilityState: getAllByA11yState,
    queryByAccessibilityState: queryByA11yState,
    queryAllByAccessibilityState: queryAllByA11yState,
    findByAccessibilityState: findByA11yState,
    findAllByAccessibilityState: findAllByA11yState
  };
};

exports.bindByA11yStateQueries = bindByA11yStateQueries;
