// @testing-library/react-native/build/queries/labelText.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.bindByLabelTextQueries = void 0;

var _matchStringProp = require("../helpers/matchers/matchStringProp");

var _makeQueries = require("./makeQueries");

const queryAllByLabelText = instance => function queryAllByLabelTextFn(text) {
  return instance.findAll(node => typeof node.type === 'string' && (0, _matchStringProp.matchStringProp)(node.props.accessibilityLabel, text));
};

const getMultipleError = labelText => `Found multiple elements with accessibilityLabel: ${String(labelText)} `;

const getMissingError = labelText => `Unable to find an element with accessibilityLabel: ${String(labelText)}`;

const {
  getBy,
  getAllBy,
  queryBy,
  queryAllBy,
  findBy,
  findAllBy
} = (0, _makeQueries.makeQueries)(queryAllByLabelText, getMissingError, getMultipleError);

const bindByLabelTextQueries = instance => ({
  getByLabelText: getBy(instance),
  getAllByLabelText: getAllBy(instance),
  queryByLabelText: queryBy(instance),
  queryAllByLabelText: queryAllBy(instance),
  findByLabelText: findBy(instance),
  findAllByLabelText: findAllBy(instance)
});

exports.bindByLabelTextQueries = bindByLabelTextQueries;
