// @testing-library/react-native/build/queries/makeQueries.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.makeQueries = makeQueries;

var _errors = require("../helpers/errors");

var _waitFor = _interopRequireDefault(require("../waitFor"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function makeQueries(queryAllByQuery, getMissingError, getMultipleError) {
  function getAllByQuery(instance) {
    return function getAllFn(predicate, options) {
      const results = queryAllByQuery(instance)(predicate, options);

      if (results.length === 0) {
        throw new _errors.ErrorWithStack(getMissingError(predicate), getAllFn);
      }

      return results;
    };
  }

  function queryByQuery(instance) {
    return function singleQueryFn(predicate, options) {
      const results = queryAllByQuery(instance)(predicate, options);

      if (results.length > 1) {
        throw new _errors.ErrorWithStack(getMultipleError(predicate), singleQueryFn);
      }

      if (results.length === 0) {
        return null;
      }

      return results[0];
    };
  }

  function getByQuery(instance) {
    return function getFn(predicate, options) {
      const results = queryAllByQuery(instance)(predicate, options);

      if (results.length > 1) {
        throw new _errors.ErrorWithStack(getMultipleError(predicate), getFn);
      }

      if (results.length === 0) {
        throw new _errors.ErrorWithStack(getMissingError(predicate), getFn);
      }

      return results[0];
    };
  }

  function findAllByQuery(instance) {
    return function findAllFn(predicate, queryOptions, waitForOptions) {
      return (0, _waitFor.default)(() => getAllByQuery(instance)(predicate, queryOptions), waitForOptions);
    };
  }

  function findByQuery(instance) {
    return function findFn(predicate, queryOptions, waitForOptions) {
      return (0, _waitFor.default)(() => getByQuery(instance)(predicate, queryOptions), waitForOptions);
    };
  }

  return {
    getBy: getByQuery,
    getAllBy: getAllByQuery,
    queryBy: queryByQuery,
    queryAllBy: queryAllByQuery,
    findBy: findByQuery,
    findAllBy: findAllByQuery
  };
}
